File: 10-code-control/16-percentage-rollout/python/16-percentage-rollout.py
#!/usr/bin/env python3
"""Serve the static percentage-rollout grid navigator."""

from __future__ import annotations

import json
import os
import sys
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

import ldclient
from ldclient import Config
from ldclient.client import LDClient

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from rollout import evaluate_rollout, normalize_username  # noqa: E402

ROOT = Path(__file__).parent
_ld_client: LDClient | None = None


def init_launchdarkly() -> None:
    """Initialize one process-wide SDK client for percentage evaluation."""
    global _ld_client
    sdk_key = (os.environ.get("LD_SDK_KEY") or "").strip()
    if not sdk_key:
        print("Warning: LD_SDK_KEY not set — highlight defaults to none.", flush=True)
        return
    ldclient.set_config(Config(sdk_key))
    _ld_client = ldclient.get()
    if not _ld_client.is_initialized():
        print("Warning: LaunchDarkly SDK did not initialize.", flush=True)


def json_response(
    handler: SimpleHTTPRequestHandler, status: int, payload: object
) -> None:
    """Write a no-cache JSON response."""
    body = json.dumps(payload).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Content-Length", str(len(body)))
    handler.send_header("Cache-Control", "no-store")
    handler.end_headers()
    handler.wfile.write(body)


class Handler(SimpleHTTPRequestHandler):
    """Serve the UI and evaluate the rollout for the supplied username key."""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(ROOT), **kwargs)

    def do_GET(self) -> None:
        parsed = urlparse(self.path)
        if parsed.path == "/api/flags":
            params = parse_qs(parsed.query, keep_blank_values=True)
            try:
                username = normalize_username((params.get("username") or [""])[0])
                json_response(self, 200, evaluate_rollout(_ld_client, username))
            except ValueError as exc:
                json_response(self, 400, {"error": str(exc)})
            return
        if parsed.path == "/api/bootstrap":
            json_response(
                self,
                200,
                {
                    "appBanner": "16-percentage-rollout[python]",
                    "flagKey": "enable-grid-selection-highlight-pct",
                    "rollout": "30% green / 70% none",
                    "port": int(os.environ.get("PORT") or "8160"),
                },
            )
            return
        super().do_GET()


def main() -> None:
    """Start the web server and close the SDK during shutdown."""
    init_launchdarkly()
    port = int(os.environ.get("PORT") or "8160")
    server = ThreadingHTTPServer(("127.0.0.1", port), Handler)
    print("16-percentage-rollout[python]")
    print(f"Open http://127.0.0.1:{port}/")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()
        if _ld_client is not None:
            _ld_client.close()


if __name__ == "__main__":
    main()

File: 10-code-control/16-percentage-rollout/node/16-percentage-rollout.js
#!/usr/bin/env node
/** Serve the static percentage-rollout grid navigator. */

const http = require("http");
const fs = require("fs");
const path = require("path");
const LaunchDarkly = require("@launchdarkly/node-server-sdk");
const { evaluateRollout, normalizeUsername } = require("../rollout");

const PORT = Number(process.env.PORT || 8080);
const ROOT = __dirname;
let ldClient = null;

async function initLaunchDarkly() {
  const sdkKey = (process.env.LD_SDK_KEY || "").trim();
  if (!sdkKey) {
    console.warn("Warning: LD_SDK_KEY not set — highlight defaults to none.");
    return;
  }
  ldClient = LaunchDarkly.init(sdkKey);
  try {
    await ldClient.waitForInitialization({ timeout: 5 });
  } catch {
    console.warn("Warning: LaunchDarkly SDK did not initialize.");
    ldClient.close();
    ldClient = null;
  }
}

function sendJson(res, status, body) {
  const payload = JSON.stringify(body);
  res.writeHead(status, {
    "Content-Type": "application/json; charset=utf-8",
    "Content-Length": Buffer.byteLength(payload),
    "Cache-Control": "no-store",
  });
  res.end(payload);
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url, `http://127.0.0.1:${PORT}`);
  if (req.method === "GET" && url.pathname === "/api/flags") {
    try {
      const username = normalizeUsername(url.searchParams.get("username"));
      return sendJson(res, 200, await evaluateRollout(ldClient, username));
    } catch (error) {
      return sendJson(res, 400, { error: error.message });
    }
  }
  if (req.method === "GET" && url.pathname === "/api/bootstrap") {
    return sendJson(res, 200, {
      appBanner: "16-percentage-rollout[node]",
      flagKey: "enable-grid-selection-highlight-pct",
      rollout: "30% green / 70% none",
      port: PORT,
    });
  }

  const urlPath = url.pathname === "/" ? "/index.html" : url.pathname;
  const filePath = path.resolve(ROOT, `.${urlPath}`);
  if (!filePath.startsWith(`${ROOT}${path.sep}`)) {
    res.writeHead(403);
    return res.end("Forbidden");
  }
  fs.readFile(filePath, (error, data) => {
    if (error) {
      res.writeHead(404);
      return res.end("Not found");
    }
    res.writeHead(200, {
      "Content-Type": filePath.endsWith(".html") ? "text/html" : "text/plain",
    });
    res.end(data);
  });
});

initLaunchDarkly().then(() => {
  server.listen(PORT, "127.0.0.1", () => {
    console.log("16-percentage-rollout[node]");
    console.log(`Open http://127.0.0.1:${PORT}/`);
  });
});

process.on("SIGINT", () => {
  if (ldClient) ldClient.close();
  process.exit(0);
});
